Modules/atexitmodule.c

```diff
@@ -89,23 +70,22 @@ _PyAtExit_Init(PyInterpreterState *interp)
     // _PyAtExit_Init() must only be called once
     assert(state->callbacks == NULL);
 
-    state->callback_len = 32;
-    state->ncallbacks = 0;
-    state->callbacks = PyMem_New(atexit_py_callback*, state->callback_len);
+    state->callbacks = PyList_New(0);
     if (state->callbacks == NULL) {
+        PyErr_WriteUnraisable(NULL);
```

Comment:
It's not needed to log the exception explicitly, it's done automatically if an init function fails and sets an exception.